sycl/doc/extensions/proposed/sycl_ext_oneapi_group_sort.asciidoc

```diff
@@ -0,0 +1,976 @@
+    sycl::nd_range<1>{ /*global_size = */ {1024}, /*local_size = */ {256} },
+    [=](sycl::nd_item<1> id) {
+      size_t group_id = id.get_group(0);
+      auto keys_ptr = keys_acc.get_pointer() + group_id * n;
+      auto vals_ptr = vals_acc.get_pointer() + group_id * n;
+      auto first = oneapi::dpl::make_zip_iterator(keys_ptr, vals_ptr);
+
+      my_sycl::joint_sort(
+        // create group excutor using deduction guides
+        my_sycl::group_with_scratchpad(
+          id.get_group(),
+          sycl::span{scratch.get_pointer(), temp_memory_size}
+        ),
+        first,
+        first + n,
+        [](auto x, auto y){ return std::get<0>(x) < std::get<0>(y); }
+      );
+    });
+  });
+...
+----
+
+4.Using `sort_over_group` and `radix_sorter` with fixed-size arrays for key-value sorting.
+
+[source,c++]
+----
+...
+namespace my_sycl = sycl::ext::oneapi::experimental;
+
+sycl::range<1> local_range{256};
+constexpr std::size_t ItemsPerWorkItem = 8;
+
+// predefine radix_sorter to calculate local memory size
+using RSorter = my_sycl::radix_sorter<T, my_sycl::sorting_order::descending>;
+// calculate required local memory size
+size_t temp_memory_size =
+    RSorter::key_value_memory_required(sycl::memory_scope::work_group, local_range);
+
+q.submit([&](sycl::handler& h) {
+  auto keys_acc = sycl::accessor(keys_buf, h);
+  auto vals_acc = sycl::accessor(vals_buf, h);
+  auto scratch = sycl::local_accessor<std::byte, 1>( {temp_memory_size}, h);
+
+  h.parallel_for(
+    sycl::nd_range<1>{ local_range, local_range },
+    [=](sycl::nd_item<1> id) {
+
+      T keys_private[ItemsPerWorkItem];
+      T vals_private[ItemsPerWorkItem];
+      auto idx = id.get_global_id();
+      for(std::size_t i = 0; i < ItemsPerWorkItem; ++i )
+      {
+        keys_private[i] = keys_acc[idx * ItemsPerWorkItem + i];
+        vals_private[i] = vals_acc[idx * ItemsPerWorkItem + i];
+      }
+
+      my_sycl::sort_over_group(
+        id.get_group(),
+        sycl::span{keys_private},
+        sycl::span{vals_private},
+        RSorter(sycl::span{scratch.get_pointer(), temp_memory_size})
+      );
+      ...
+    });
+  });
+...
+----
+
+== Issues for later investigations
+
+. It can be a separate proposal for key-value sorting basing on Projections.
+It needs to be investigated what is the response for that.
+. Sorter traits can be useful if there are Finder, Reducer or other objects
+will be added to the Spec to be used with other Group algorithms, e.g. find, reduce.
+. Predefined sorters can be revisited to find a better interfaces for `memory_required` overloads.
+. `group_with_scratchpad` can be replaced with concepts that requires 2 methods.
+. Is Sorter needed to be applied to keys only or to keys and values both in case of key-value sorting?
+. Do we need to have separate predefined sorters for fixed-size array interfaces?
+e.g. instead of changing `default_sorter` and `radix_sorter` to have new sorters `default_span_sorter`, `radix_span_sorter`.

```

Comment:
I think this would make the API less confusing.  As it is now, it seems confusing to have two `memory_required` functions  for `default_sorter`.

Reply:
2 `memory_required` functions are due to `joint_sort` and `sort_over_group` differences. There is a point in this list to find a better interfaces for `memory_required` overloads. Another thing is that sorters are differentiated by algorithms, not by data storages they accept as parameters now.

Reply:
I looked at this some more, and I still have the same concern.  The `default_sorter` has four overloads of `operator()` and four functions that are named like `memory_required`.  This doesn't seem scalable to me as we add new variations of sorting. I also think it will be easy for applications to mistakenly call the wrong overload of `memory_required`, and this will lead to hard-to-debug problems.

I think it would be better to have one class for each type of predefined sorter.  They could be named like this:

* `sorter::joint_sort`
* `sorter::sort_over_group`
* `sorter::sort_over_group_span`
* `sorter::sort_over_group_span_key_value`
* `sorter::joint_sort_radix`
* `sorter::sort_over_group_radix`
* `sorter::sort_over_group_radix_span`
* `sorter::sort_over_group_radix_span_key_value`

Each of those classes would have one constructor, one `operator()`, and one function named `memory_required`.  This seems a lot more scalable as we add more sorting variations, and there will be no confusion about which `memory_required` function to call.

I'm pretty open on the  names, but I think the name needs to make it clear where the sorter can be used.  I thought it would be nice to put all the predefined sorters in their own namespace, which is what I show above, but I'm not wedded to that.

If we go with the naming I propose above, we should change the name of the section "Functions with fixed-size arrays" to something like "Sorting functions over spans".  This ties the name of the predefined sorters (e.g. `sort_over_group_span` to the relevant section in the spec.

Reply:
@Pennycook, do you remember why we decided to go with less sorters incapsulating all `operator()` in sorters? Only to not multiply entities? Unfortunately, I don't remember all arguments.

Reply:
I don't remember, but I _think_ it might have had something to do with wanting to avoid users providing the same information multiple times.  Developers already have to specify what kind of sort they want by calling a different function; if there was a different sorter as well, there'd be duplication:

```c++
sycl::joint_sort(g, first, last, sorter::joint_sort());
```
vs
```c++
sycl::joint_sort(g, first, last, sorter());
```

A developer never needs to call `operator()` for a sorter directly, so I personally don't mind that there are lots of overloads.

But I do agree with @gmlueck's concern that it's very easy to call the wrong version of `memory_required`.  Could we do something here with tag types, or other arguments?  For example, we could define the operator overloads as:

```c++
template <typename Group, typename Ptr>
void operator()(sorter::joint_sort_tag_t, Group g, Ptr first, Ptr last);

template<typename Group, typename T>
T operator()(sorter::sort_over_group_tag_t, Group g, T val);

template<typename Group, typename K, typename V>
T operator()(sorter::sort_over_group_tag_t, Group g, K key, V val);
```

And define each `memory_required` to (roughly) match that operator, e.g.:

```c++
template <typename Ptr>
std::size_t memory_required(sorter::joint_sort_tag_t, sycl::memory_scope scope, Ptr);

template<typename T>
std::size_t memory_required(sorter::sort_over_group_tag_t, sycl::memory_scope scope, T);

template<typename K, typename V>
std::size_t memory_required(sorter::sort_over_group_tag_t, sycl::memory_scope scope, K, V);
```

I'd prefer using overloading to distinguish between value, span and {key, value} mainly to keep the number of tags small.  But I don't feel strongly about that.  Also, the sketch above was based on my assumption it would be easier to remember which overload to call if a developer could write something like:

```c++
auto mem = memory_required(sorter::sort_over_group_tag_t, sub_group::fence_scope, float{}, int{});
```

...but I'm not opposed to requiring templates.  My sketch might require all types that can be sorted to be default constructible, which we might not want.

Reply:
Let's look at a more complete example to compare the two styles.  Here is the version with a tag-type (using the default sorter and the fixed-array sort):

```
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

size_t temp_size = default_sorter::memory_required<int, NUM>(
  sorter::sort_over_group_span_tag_t,
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      default_sorter{span{sratch.get_pointer(), temp_size});
  });
});
```

And here it is using different classes for the predefined sorters:

```
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

using sorter = sorter::sort_over_group_span;
size_t temp_size = sorter::memory_required<int, NUM>(
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      sorter{span{sratch.get_pointer(), temp_size});
  });
});
```

The approach with the separate classes doesn't lead to any duplication of code, and the code is no more verbose.  In fact, it allows you to have a single `using` statement which selects the predefined sorter, and this `using` statement keeps the call to `memory_required` in sync with the instantiation of the sorter object in the kernel.

Another thing I like is that the separate classes allow us to name the "memory required" function the same thing in all cases (`memory_required`).  By contrast, the existing style requires one of the functions to be named `key_value_memory_required` because it would otherwise lead to two overloads with the same signature.

Reply:
Thanks for the complete example.  I wouldn't have thought of using a `using` statement like that, and I agree that makes the examples much nicer.  I like the style you're suggesting, and you've convinced me types are a better solution than tag types.

Bikeshedding: I think we should look to see if there's anything we can do to reduce the number of top-level types.  For example, we could have a top-level class that defines types for each of the four variants:
- `default_sorter::joint_sort`
- `default_sorter::sort_over_group`
- `default_sorter::sort_over_group_span`
- `default_sorter::sort_over_group_span_key_value`
- `radix_sorter::joint_sort`
- `radix_sorter::sort_over_group`
- `radix_sorter::sort_over_group_span`
- `radix_sorter::sort_over_group_span_key_value`

That would make it simpler to pass around the entire "family" of related sorters, while still retaining the benefits you outlined. This would make it possible for developers to write functions/libraries that accept a sorter, and decide whether to call `joint_sort` or `sort_over_group` inside the function/library.

We could also choose to allow code like the below the compile in addition to the `using` version:

```c++
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

size_t temp_size = default_sorter::sort_over_group_span::memory_required<int, NUM>(
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      default_sorter{span{scratch.get_pointer(), temp_size}); // forwards to `default_sorter::sort_over_group_span`.
  });
});
```

I don't know what names we'd use to distinguish between the top-level class and the individual classes.

Reply:
I assume you are proposing an organization like this:

```
struct default_sorter {
  template<std::size_t Extent>
  default_sorter(sycl::span<std::byte, Extent> scratch);  // Forwards to one of the constructors below

  class joint_sort {
   public:
    template<std::size_t Extent>
    joint_sort(sycl::span<std::byte, Extent> scratch);

    template<typename Group, typename Ptr>
    void operator()(Group g, Ptr first, Ptr last);

    template<typename T>
    static constexpr std::size_t
    memory_required(sycl::memory_scope scope, std::size_t range_size);
  };

  class sort_over_group {
   public:
    template<std::size_t Extent>
    sort_over_group(sycl::span<std::byte, Extent> scratch);

    template<typename Group, typename T>
    T operator()(Group g, T val);

    template<typename T, int dimensions>
    static constexpr std::size_t
    memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range);
  };

  /* ... */
};
```

However, I don't see how that can work because the `default_sorter` constructor won't know which sub-class constructor to call.  All the sub-class constructors have the same signature.

I suppose you could solve this by adding a tag-type parameter to the `default_sorter` constructor.  Somehow it seems weird, though, to have both a tag-type and also a separate structure for each sorter variation.  Do we gain enough from  the nested-structure approach to make this worthwhile?

Reply:
> Aside: While sketching this I noticed something a bit weird, and I can't tell if it's an issue. If the <std::size_t Extent> template is on the constructor and not on default_sorter itself, how does the sorter store the span?

This is a good observation.  I looked at the existing implementation, and it seems that the scratch space size is stored as a regular member variable (not a `constexpr` quantity).  Reading the spec more closely, I think it is possible for `Extent` to be `std::dynamic_extent`.

> I was envisioning that the forwarding would happen inside of sort_over_group. No tag type is required, because the function already knows what kind of sort it's doing:

Thanks for the details.  Yes, that looks like it could work.  I think @andreyfe1 is on vacation now, so we'll have to wait for him to return to get his opinion.

The organization you propose would allow either of the following usage models:

```
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

using sorter = default_sorter::sort_over_group_span;
size_t temp_size = sorter::memory_required<int, N>(
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      sorter{span{sratch.get_pointer(), temp_size});
  });
});
```

Or:

```
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

size_t temp_size = default_sorter::sort_over_group_span::memory_required<int, N>(
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      default_sorter{span{sratch.get_pointer(), temp_size});
  });
});
```

Personally, I still like the first model better because the `using` statement keeps the `memory_required` call in sync with the use of the sorter.  If they get out-of-sync, the compiler will raise a compile-time error.  The second model does not raise a compile-time error if they get out of sync, though it does have one less line.

If we go with this style, I think we should illustrate both models with examples.

Reply:
@Pennycook, @gmlueck, thanks for the fruitful discussion. The style looks good to me. Let me prototype this approach

Reply:
>  If the <std::size_t Extent> template is on the constructor and not on default_sorter itself, how does the sorter store the span?

It's stored as pointer and integer value (2 fields).

> I was envisioning that the forwarding would happen inside of sort_over_group. No tag type is required, because the function already knows what kind of sort it's doing:

I tried to prototype the approach where `default_sorter` can be constructed from scratch and others, not `default_sorter::sort_over_group`. For that we will need the following:
```cpp
// (1)
template<typename Compare>
struct default_sorter {
  template<std::size_t Extent>
  default_sorter(sycl::span<std::byte, Extent> scratch, Compare compare);  // Forwards to one of the constructors below
  
  Compare get_comparator() const;
  /**/ get_memory() const;

  class sort_over_group {
   public:
    template<std::size_t Extent>
    sort_over_group(sycl::span<std::byte, Extent> scratch);

    template<typename Group, typename T>
    T operator()(Group g, T val);

    template<typename T, int dimensions>
    static constexpr std::size_t
    memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range);
  };

  /* ... */
};
```
So, we will need an additional constructor and getters to let `set_over_group` function to create an instance of internal class later.

In my opinion it complicates the specification more than usability that it gives. If we have `default_sorter` as a container for other classes it will look like following:
```cpp
// (2)
template<typename Compare>
struct default_sorter {

  class sort_over_group {
   public:
    template<std::size_t Extent>
    sort_over_group(sycl::span<std::byte, Extent> scratch);

    template<typename Group, typename T>
    T operator()(Group g, T val);

    template<typename T, int dimensions>
    static constexpr std::size_t
    memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range);
  };

  /* ... */
};
```
and users can do something like @gmlueck mentioned above:
```cpp
constexpr int N = /* elements per work item */;
constexpr int R = /* local range size */;

using sorter = default_sorter::sort_over_group_span;
size_t temp_size = sorter::memory_required<int, N>(
  memory_scope::work_group,
  {R});

q.submit([&](sycl::handler& cgh) {
  cgh.parallel_for(/*...*/, [=](nd_item<1> id) {
    float vals[N];
    sort_over_group(
      id.get_group(),
      sycl::span{vals},
      sorter{span{sratch.get_pointer(), temp_size});
  });
});
```
@Pennycook, what do you think if we go with the scheme (2)?

Reply:
> It's stored as pointer and integer value (2 fields).

That prevents any sort of specialization for dynamic vs non-dynamic extents, right?  I think this is probably fine -- since the span is just representing the amount of scratch memory and it might be hard to optimize anything based on that -- but I just want to make sure.

> I tried to prototype the approach where `default_sorter` can be constructed from scratch and others, not `default_sorter::sort_over_group`. For that we will need the following:

Thanks for prototyping.  If it will be cleaner, I'm okay with (2).

One thing that occurred to me while looking through your snippets, though, is that there may be a third alternative.  Sorry to both you and @gmlueck for doing this (again!)...

Both (1) and (2) still allow mistakes to be made, because the developer can call `memory_required` with a different type to the one that ends up being sorted.  Your third snippet makes this mistake, I think -- the size of the temporary comes from `sorter::memory_required<int, ...>` but the values being sorted are `float`s. I appreciate that in this particular case both types are the same size, but you could have made the same mistake with different types.

That makes me wonder if it wouldn't be better to move the type information onto a template of the nested class, like so:

```c++
// (3)
template<typename Compare>
struct default_sorter {

  template <typename T>
  class sort_over_group {
   public:
    template<std::size_t Extent>
    sort_over_group(sycl::span<std::byte, Extent> scratch, Compare comparator);

    template<typename Group>
    T operator()(Group g, T val);

    template<int dimensions>
    static constexpr std::size_t
    memory_required(sycl::memory_scope scope, sycl::range<dimensions> local_range);
  };

  /* ... */
};
```

Since the nested class is now templated, we could choose to use specialization instead of multiple names to handle the three different cases:
- `sort_over_group` => `sort_over_group<T, void>` (`void` for exposition, it could be defaulted)
- `sort_over_group_span` => `sort_over_group<sycl::span<T>, void>` (ditto)
- `sort_over_group_key_value` => `sort_over_group<K, V>`

...but that might be overkill?

Reply:
> * sort_over_group => sort_over_group<T, void> (void for exposition, it could be defaulted)
> * sort_over_group_span => sort_over_group<sycl::span<T>, void> (ditto)
> * sort_over_group_key_value => sort_over_group<K, V>

I'm not sure about this part.  Having separate names seems clearer to me, rather than relying on different template parameters.  Also, doesn't this assume that no sorter will ever require more than two template parameters?  What if we add another sorter in the future that requires a third parameter?

At this point, the outer struct `default_sorter` doesn't contain anything other than the inner sorter classes.  Does `default_sorter` need to be a struct, or should it just be a namespace?

Reply:
> I'm not sure about this part. Having separate names seems clearer to me, rather than relying on different template parameters. Also, doesn't this assume that no sorter will ever require more than two template parameters? What if we add another sorter in the future that requires a third parameter?

Good points.  It probably is a bad idea to assume that we will only ever need two template parameters.

I'm just looking for ways to simplify because the group algorithms already have lots of names (e.g. "over group" and "joint") and this extension will set a precedent for other algorithms.  I'm worried that if there are too many names, the algorithms will look too complicated and people will get confused.  But I might be worrying prematurely.


> At this point, the outer struct `default_sorter` doesn't contain anything other than the inner sorter classes. Does `default_sorter` need to be a struct, or should it just be a namespace?

`default_sorter` is still templated in Andrey's snippet.  But I think if we moved that template to the inner classes, a namespace would work.

Reply:
> But I think if we moved that template to the inner classes, a namespace would work.

Agree. `default_sorter` will be just a wrapper over inner classes. No additional fields/methods are needed for this

Reply:
I've finalized with prototyping. With the last changes we will have 2 families of sorters: default sorters and radix sorters:
```cpp

  namespace default_sorters {

    template<typename Compare = std::less<>>
    class joint_sort_sorter{
    public:
        template<std::size_t Extent>
        joint_sort_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group, typename Ptr>
        void operator()(Group g, Ptr begin, Ptr end);

        template<typename T>
        static constexpr size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename Compare = std::less<>>
    class sort_over_group_sorter{
    public:
        template<std::size_t Extent>
        sort_over_group_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group, typename T>
        T operator()(Group g, T val);

        template<typename T>
        static constexpr size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename Compare = std::less<>>
    class sort_over_group_span_sorter{
    public:
        template<std::size_t Extent>
        sort_over_group_span_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group, typename T, std::size_t ItemsPerWorkItem, typename Properties>
        void operator()(Group g, sycl::span<T, ItemsPerWorkItem> values, Properties properties);

        template<typename T, std::size_t ItemsPerWorkItem>
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename Compare = std::less<>>
    class sort_over_group_span_key_value_sorter{
    public:
        template<std::size_t Extent>
        sort_over_group_span_key_value_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group, typename T, typename U, std::size_t ItemsPerWorkItem, typename Properties>
        void operator()(Group g, sycl::span<T, ItemsPerWorkItem> keys, sycl::span<U, ItemsPerWorkItem> values, Properties properties)

        template<typename T, typename U, std::size_t ItemsPerWorkItem>
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };
  }

  namespace radix_sorters{

    template<typename T, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class joint_sort_sorter
    {
    public:
        template<std::size_t Extent>
        joint_sort_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<typename Group, typename Ptr>
        void
        operator()(Group g, Ptr begin, Ptr end);

        // memory_helpers
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class sort_over_group_sorter
    {
    public:
        template<std::size_t Extent>
        sort_over_group_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<typename Group, typename T>
        T operator()(Group g, T val);

        // memory_helpers
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class sort_over_group_span_sorter
    {
    public:
        template<std::size_t Extent>
        sort_over_group_span_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<std::size_t ItemsPerWorkItem, typename Group, typename Properties>
        void
        operator()(Group g, sycl::span<T, ItemsPerWorkItem> values, Properties properties);

        // memory_helpers
        template<std::size_t ItemsPerWorkItem>
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class sort_over_group_span_key_value_sorter
    {
    public:
        template<std::size_t Extent>
        sort_over_group_span_key_value_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<std::size_t ItemsPerWorkItem, typename U, typename Group, typename Properties>
        void
        operator()(Group g, sycl::span<T, ItemsPerWorkItem> keys, sycl::span<U, ItemsPerWorkItem> values, Properties properties);

        // memory_helpers
        template<typename U, std::size_t ItemsPerWorkItem>
        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

}
```
Names can be changed. Long lines will be split.
@Pennycook, @gmlueck, what do you think about it?

Reply:
There are a few instances where you still have `typename T` on some of the member functions in the default sorters; I suspect this is a typo, but just to be clear -- I think those should be moved to the first template parameter of the sorter class, to make things less error-prone:

```c++
template <typename T, typename Compare = std::less<>>
class sort_over_group_sorter{
public:
    template<std::size_t Extent>
    sort_over_group_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

    template<typename Group>
    T operator()(Group g, T val);

    static constexpr size_t
    memory_required(sycl::memory_scope scope, std::size_t range_size);
};
```

There are also some combinations missing (like key_value without span) but I assume that's to keep things readable?

I really hope that we can come up with better names, because `sort_over_group_span_key_value_sorter` is (imho) far too long.  The fact it contains both "sort" and "sorter" feels redundant.  If we keep this arrangement of things, my suggestion would be: `group_sorter`, `group_sorter_span_kv`, `group_sorter_kv`, etc and `joint_sorter`, `joint_sorter_span_kv`, `joint_sorter_kv`, etc

I appreciate my last proposal to have everything in one class was too ambitious, but how would people feel about at least folding the "span" versions of things into one class?  That way we could have:

```c++
template <typename T, typename Compare = std::less<>>
class group_sorter{
public:
    template<std::size_t Extent>
    group_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

    template<typename Group>
    T operator()(Group g, T val);

    template<std::size_t ItemsPerWorkItem, typename Group, typename Properties>
    void operator()(Group g, sycl::span<T, ItemsPerWorkItem> values, Properties properties);

    // or templated on ItemsPerWorkItem = 1
    static constexpr size_t
    memory_required(sycl::memory_scope scope, std::size_t range_size, std::size_t per_work_item = 1);
};
```

...and we'd only need two names: `group_sorter` and `group_sorter_kv`.

I guess what I'm really asking is -- do we really need the complexity and cognitive burden of two classes if the only difference between the `T` and `span<T>` versions is the number of things being sorted?  I'd really like if we could do this, since it would reduce the number of unique names we need by 2x.

Reply:
I agree with @Pennycook about moving the `typename T` and `typename U` to the class.  I think we could also move the `ElementsPerWorkItem` template parameter to the class.

Reply:
> I think we could also move the `ElementsPerWorkItem` template parameter to the class.

Ooh, yeah, that's an even better idea.  Then we can make the `operator()` that accepts a single value available only if `NumWorkItems` == 1.

Reply:
And we can keep the other `operator()` in sync with the call to `memory_required`.

Reply:
Looks interesting. In my version above I thought that by "sorter" we could name a class that encapsulated one `operator()` and one `memory_required` function. However, if we rethink that sorter can encapsulate `operator()` for one element and/or for span over values, it also will work. As previously users can define only one `operator()` (for example, with "span") if they don't need more versions of `operator()`. Let me summarize, what is suggested: we will have 3 type of sorters: "group_sorter", "group_sorter_key_value" or "group_key_value_sorter", "joint_sorter", correct?
Something like this?
```cpp
template <typename T, std::size_t ItemsPerWorkItem = 1, typename Compare = std::less<>>
class group_sorter{
public:
    template<typename Group>
    T operator()(Group g, T val);

    template<std::size_t ItemsPerWorkItem, typename Group, typename Properties>
    void operator()(Group g, sycl::span<T, ItemsPerWorkItem> values, Properties properties);
...
};

template <typename T, typename U, typename Compare = std::less<>>
class group_key_value_sorter{
public:
    template<typename Group>
    std::tuple<T, U> operator()(Group g, T key, U value);

    template<std::size_t ItemsPerWorkItem, typename Group, typename Properties>
    void operator()(Group g, sycl::span<T, ItemsPerWorkItem> keys, sycl::span<U, ItemsPerWorkItem> values, Properties properties);
...
};

template <typename Ptr, typename Compare = std::less<>>
class joint_sorter{
public:
    template<typename Group>
    void operator()(Group g, Ptr first, Ptr last);
    
    // span version is not needed
...
};

// key-value version of joint sorter can be expressed through joint_sorter with dpl::zip_iterator
```

Regarding the concept of sorter. We can say users that Sorter can contain one or more `operator()` functions.

Reply:
Regarding following
```cpp
template<typename Group>
std::tuple<T, U> operator()(Group g, T key, U value);
```

As for me it looks natural if we say that key-value version for `sort_over_group` can be expressed through `T operator()(Group g, T val);` (`T` is a `tuple<Key, Value>`) that is in `group_sorter`, but the only problem is `radix_sorter`. For `radix_sorter` we couldn't say that `T` is a `tuple<Key, Value>` because we say that `T` for `radix_sorter` should be an arithmetic type. So, I see 2 opitions:
1. Say in the Spec that `T` can be a `tuple<Key, Value>` and for `radix_sorter` this is the special exceptions that we allow arithmetic types and tuple of 2 arithmetic types; `operator>` or `<` will be applied to "keys" only (1st component of tuple).
2. Multiply entities by adding `std::tuple<T, U> operator()(Group g, T key, U value);`
Any opinions? 2nd option looks less complicated for users. If they don't need it, they won't use this overload

Reply:
>  Let me summarize, what is suggested: we will have 3 type of sorters: "group_sorter", "group_sorter_key_value" or "group_key_value_sorter", "joint_sorter", correct?
Something like this?

I hadn't realized we could drop span and key_value from joint_sorter like this.  I think this looks very good.

To clarify an earlier comment: I'd expect the real implementation to specialize the implementation of `group_sorter` based on the `ItemsPerWorkItem` template parameter, so that `T operator()(Group g, T val);` is only present in the case of one input.  It shouldn't be possible to create a sorter for a large span and accidentally invoke the scalar operator.

Suggestion: Rename `ItemsPerWorkItem` to `InputsPerWorkItem` to avoid mixing two meanings of "Item".


> Any opinions? 2nd option looks less complicated for users. If they don't need it, they won't use this overload

I don't feel strongly about this, because the user already has to opt-in to a key-value sort using a different top-level sorter.  But I'm in favor of the 2nd option, because it's aligned with the overloads of `sort_over_group` elsewhere in the proposal:

```c++
  template<typename Group, typename T, typename U,
           std::size_t ElementsPerWorkItem,
           typename Sorter,
           typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(Group g, sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Sorter sorter,
                       Properties properties = {});
```

Keys and values as separate arguments makes it clear that it's a key-value sort, and not a sort of the tuples.  In the general case, a sort over tuples could use a `Compare` object that doesn't consider all elements of the tuple.

Reply:
> Suggestion: Rename ItemsPerWorkItem to InputsPerWorkItem to avoid mixing two meanings of "Item".

I think `ElementsPerWorkItem` that we already have for the spec is also applicable, right?

Reply:
> I think `ElementsPerWorkItem` that we already have for the spec is also applicable, right?

Yes.  Sorry, I was looking at your snippet and not at the spec.

Reply:
Finally we will have such classes:
```cpp
  namespace default_sorters {

    template<typename Compare = std::less<>>
    class joint_sorter{
    public:
        template<std::size_t Extent>
        joint_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group, typename Ptr>
        void operator()(Group g, Ptr first, Ptr last);

        template<typename T>
        static constexpr size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T,
             typename Compare = std::less<>,
             std::size_t ElementsPerWorkItem = 1>
    class group_sorter{
    public:
        template<std::size_t Extent>
        group_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group>
        T operator()(Group g, T val);

        template<typename Group, typename Properties>
        void operator()(Group g,
                        sycl::span<T, ElementsPerWorkItem> vals,
                        Properties properties);

        static constexpr size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T,
             typename U,
             typename Compare = std::less<>,
             std::size_t ElementsPerWorkItem = 1>
    class group_key_value_sorter{
    public:
        template<std::size_t Extent>
        group_key_value_sorter(sycl::span<std::byte, Extent> scratch, Compare comp = {});

        template<typename Group>
        std::tuple<T, U> operator()(Group g, T key, U value);

        template<typename Group, typename Properties>
        void
        operator()(Group g,
                   sycl::span<T, ElementsPerWorkItem> keys,
                   sycl::span<U, ElementsPerWorkItem> values,
                   Properties property);

        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };
  }

  namespace radix_sorters {

    template<typename T, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class joint_sorter
    {
    public:
        template<std::size_t Extent>
        joint_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<typename Group, typename Ptr>
        void
        operator()(Group g, Ptr begin, Ptr end);

        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T, std::size_t ElementsPerWorkItem = 1, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class group_sorter
    {
    public:
        template<std::size_t Extent>
        group_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<typename Group>
        T operator()(Group g, T val);

        template<typename Group, typename Properties>
        void operator()(Group g,
                        sycl::span<T, ElementsPerWorkItem> values,
                        Properties properties);

        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };

    template<typename T, typename U, std::size_t ElementsPerWorkItem = 1, sorting_order Order = sorting_order::ascending, unsigned int BitsPerPass = 4>
    class group_key_value_sorter
    {
    public:
        template<std::size_t Extent>
        group_key_value_sorter(sycl::span<std::byte, Extent> scratch, const std::bitset<sizeof(T) * CHAR_BIT> mask =
                        std::bitset<sizeof(T) * CHAR_BIT> (std::numeric_limits<unsigned long long>::max()));

        template<typename Group>
        std::tuple<T, U> operator()(Group g, T key, U value);

        template<typename Group, typename Properties>
        void operator()(Group g,
                        sycl::span<T, ElementsPerWorkItem> keys,
                        sycl::span<U, ElementsPerWorkItem> values,
                        Properties properties);

        static constexpr std::size_t
        memory_required(sycl::memory_scope scope, std::size_t range_size);
    };
}
```

Now let me show what we have with sorting functions:
```cpp
// 1. joint_sort
  template <typename GroupHelper, typename Ptr>
  void joint_sort(GroupHelper exec, Ptr first, Ptr last);

  template <typename GroupHelper, typename Ptr, typename Compare>
  void joint_sort(GroupHelper exec, Ptr first, Ptr last, Compare comp);

  template <typename Group, typename Ptr, typename Sorter>
  void joint_sort(Group g, Ptr first, Ptr last, Sorter sorter);

// 2. sort_over_group
  template <typename GroupHelper, typename T>
  T sort_over_group(GroupHelper exec, T value);

  template <typename GroupHelper, typename T, typename Compare>
  T sort_over_group(GroupHelper exec, T value, Compare comp);

  template <typename Group, typename T, typename Sorter>
  T sort_over_group(Group g, T value, Sorter sorter);

// 3. sort_over_group with span
  template <typename GroupHelper, typename T,
            std::size_t ElementsPerWorkItem,
            typename Properties =  ext::oneapi::experimental::properties<>>
  void sort_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> values,
                       Properties properties = {});

  template <typename GroupHelper, typename T,
            std::size_t ElementsPerWorkItem, typename Compare,
            typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> values,
                       Compare comp,
                       Properties properties = {});

  template<typename Group, typename T, std::size_t ElementsPerWorkItem,
           typename Sorter,
           typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(Group g,
                       sycl::span<T, ElementsPerWorkItem> values,
                       Sorter sorter,
                       Properties properties = {});

// 4. sort_key_value_over_group
  template <typename GroupHelper, typename T, typename U>
  T sort_key_value_over_group(GroupHelper exec, T key, U value);

  template <typename GroupHelper, typename T, typename U, typename Compare>
  T sort_over_group(GroupHelper exec, T key, U value, Compare comp);

  template <typename Group, typename T, typename U, typename Sorter>
  T sort_over_group(Group g, T key, U value, Sorter sorter);

// 5. sort_key_value_over_group with span
  template <typename GroupHelper, typename T, typename U,
            std::size_t ElementsPerWorkItem,
            typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Properties properties = {});

  template <typename GroupHelper, typename T, typename U,
            std::size_t ElementsPerWorkItem,
            typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Compare comp,
                       Properties properties = {});

  template<typename Group, typename T, typename U,
           std::size_t ElementsPerWorkItem,
           typename Sorter,
           typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(Group g, sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Sorter sorter,
                       Properties properties = {});
```

Since we decided to have 2 different types of predefined sorters (for key-value sorting and for ordinary sorting) I think it's a good idea to have 2 different types of sorting functions as well: `sort_over_group`, `sort_key_value_over_group`.

The concern I have is do we really need 3 overloads for each group of functions (3 x 5 = 15). For example, we can only have an overload with Sorter for some group, e.g. we can have only one overload for "span"-sorting functions or "key-value" sorting functions.

What is your opinion on sorting functions?

Reply:
> Since we decided to have 2 different types of predefined sorters (for key-value sorting and for ordinary sorting) I think it's a good idea to have 2 different types of sorting functions as well: sort_over_group, sort_key_value_over_group.

Agreed.

> The concern I have is do we really need 3 overloads for each group of functions (3 x 5 = 15).  For example, we can only have an overload with Sorter for some group, e.g. we can have only one overload for "span"-sorting functions or "key-value" sorting functions.

It seems odd to me that we would provide overloads for some of the functions and not others.  As a developer, I would find that very confusing, especially if the documentation provided no justification as to why that was the case.

The standard algorithms have lots of overloads, so I don't think what you've sketched looks too bad.  But if you want to cut things down, I think we should find a way to reduce the number of overloads consistently across all the options, e.g.:

- We could remove the `Compare` overload and steer developers towards the `Sorter` overload instead
- We could remove the default overload and force developers to explicitly specify a comparison function (via `Compare` or `Sorter`)

...both the options above would make things slightly more verbose, and slightly less aligned with the C++ algorithms.

Reply:
> We could remove the Compare overload and steer developers towards the Sorter overload instead

I like this option since we have a Sorter that is pretty close to Comparator, but more powerful. I find it easy for users to switch to version with sorters. If there is a request from users, I think we can add versions with comparators later. So, we will have just 2 versions: with sorters and simple one with default comparators:
```cpp
// 1. joint_sort
  template <typename GroupHelper, typename Ptr>
  void joint_sort(GroupHelper exec, Ptr first, Ptr last);

  template <typename Group, typename Ptr, typename Sorter>
  void joint_sort(Group g, Ptr first, Ptr last, Sorter sorter);

// 2. sort_over_group
  template <typename GroupHelper, typename T>
  T sort_over_group(GroupHelper exec, T value);

  template <typename Group, typename T, typename Sorter>
  T sort_over_group(Group g, T value, Sorter sorter);

// 3. sort_over_group with span
  template <typename GroupHelper, typename T,
            std::size_t ElementsPerWorkItem,
            typename Properties =  ext::oneapi::experimental::properties<>>
  void sort_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> values,
                       Properties properties = {});

  template<typename Group, typename T, std::size_t ElementsPerWorkItem,
           typename Sorter,
           typename Properties = ext::oneapi::experimental::properties<>>
  void sort_over_group(Group g,
                       sycl::span<T, ElementsPerWorkItem> values,
                       Sorter sorter,
                       Properties properties = {});

// 4. sort_key_value_over_group
  template <typename GroupHelper, typename T, typename U>
  T sort_key_value_over_group(GroupHelper exec, T key, U value);

  template <typename Group, typename T, typename U, typename Sorter>
  T sort_key_value_over_group(Group g, T key, U value, Sorter sorter);

// 5. sort_key_value_over_group with span
  template <typename GroupHelper, typename T, typename U,
            std::size_t ElementsPerWorkItem,
            typename Properties = ext::oneapi::experimental::properties<>>
  void sort_key_value_over_group(GroupHelper gh,
                       sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Properties properties = {});

  template<typename Group, typename T, typename U,
           std::size_t ElementsPerWorkItem,
           typename Sorter,
           typename Properties = ext::oneapi::experimental::properties<>>
  void sort_key_value_over_group(Group g,
                       sycl::span<T, ElementsPerWorkItem> keys,
                       sycl::span<U, ElementsPerWorkItem> values,
                       Sorter sorter,
                       Properties properties = {});
```

Reply:
Sorry for ignoring this discussion for the last few days.  Is the comment "Finally we will have such classes" the latest proposal for the predefined sorters?  I have a couple minor comments:

* I assume there is a typo with `default_sorters::joint_sorter`.  The template parameter `T` should be on the class, not the `memory_required` function, right?

* Should we change the order of the template parameters in cases like this:
  ```
  template<typename T,
             typename Compare = std::less<>,
             std::size_t ElementsPerWorkItem = 1>
  ```
  Wouldn't it be more convenient if `ElementsPerWorkItem` was before `Compare`?  I presume it will be much more common to want to override the default of `ElementsPerWorkItem`.

Reply:
> Is the comment "Finally we will have such classes" the latest proposal for the predefined sorters?

At current state, yes. I don't think we need to invest a lot of time into it now to make it ideal since we need to unblock the implementation that potentially can raise more questions. I'd suggest to have it as a final version if there are no concerns that prevents us from implementing it right now. Later we can make some changes into the spec that won't impact implementation significantly (e.g. changing names, namespaces, layouts of classes, something that won't lead to big changes into implementation of the algorithm). Currently we don't have anything in the Spec that relates to "span"-versions of Group algorithms, but it's pretty useful to have it.

> The template parameter T should be on the class, not the memory_required function, right?

I don't think so. The difference is that `T` means nothing for `joint_sorter` since it operates with iterators, not the type of values `*_over_group` functions operates with.

> Wouldn't it be more convenient if ElementsPerWorkItem was before Compare?

I also thought regarding where to place this parameter. Testing different variants, `ElementsPerWorkItem` needs only for "span" versions. What if users want to use only `sort_over_group` function with 1 value? If we place  `ElementsPerWorkItem` before, they will need to mention `1` every time, but for `sort_over_group` it means nothing (looks like users want to have a "span"-version with 1 element inside the span. I found it extra for current proposal).

Reply:
> I don't think so. The difference is that `T` means nothing for `joint_sorter` since it operates with iterators, not the type of values *_over_group functions operates with.

I see your point.  However, it still seems inconsistent to place the template parameter `T` differently for this one sorter.  I think it would be easier for users to remember how to use them if the template parameters were in a consistent place for all of them.

> What if users want to use only `sort_over_group` function with 1 value? If we place `ElementsPerWorkItem` before, they will need to mention `1` every time, but for `sort_over_group` it means nothing

This may indicate that it was a bad idea to use the same default sorter class for the 1-element version and also the span version.  We could avoid this by having two classes like:

```
template<typename T,
         typename Compare = std::less<>>
class group_sorter {
};

template<typename T,
         std::size_t ElementsPerWorkItem,
         typename Compare = std::less<>>
class group_span_sorter {
};
```

It's probably also advantageous to *not* have a default value for `ElementsPerWorkItem` in the span version since most uses of this sorter will not use the value `1`.

This also avoids another concern I have.  By using a single class for both 1-element sort and span-sort, we assume that the memory required for a 1-element sort is the same as the memory required for a span sort when the span is 1 element.  This might be the case for our implementation, but I'm not sure it's good to assume it will be the case for every implementation.  Splitting this into two classes gives the implementation more freedom here.

Reply:
> However, it still seems inconsistent to place the template parameter T differently for this one sorter

You're right it's inconsistency between sorters, but this is the point: sorters are different and their usage is different. Let's imagine that users have `dpl::zip_iterator` as an input for the `joint_sort` function. In this case `T` is an `std::iterator_traits<dpl::zip_iterator<Args...>>::value_type`. And this parameter is not used. I still think that this parameter is extra for `joint_sorter`.

Reply:
> This may indicate that it was a bad idea to use the same default sorter class for the 1-element version and also the span version. It's probably also advantageous to not have a default value for ElementsPerWorkItem in the span version since most uses of this sorter will not use the value 1.
This also avoids another concern I have. By using a single class for both 1-element sort and span-sort, we assume that the memory required for a 1-element sort is the same as the memory required for a span sort when the span is 1 element. This might be the case for our implementation, but I'm not sure it's good to assume it will be the case for every implementation. Splitting this into two classes gives the implementation more freedom here

Well, we have already been there recently. I suppose this comment doesn't block us from merging the proposal, right? I'm afraid that this review takes too much time now and I'm also afraid that it won't be implemented in the coming future despite the priorities that we have. We have a lot of work besides spec. I'd suggest to not be ideal in all details and finalize what we can merge now. As I mentioned we can continue polishing it later unblocking many other processes we need.

Reply:
Changes are available with the latest commit